lib/executor/variable_looping_vus.go

```diff
@@ -636,40 +633,88 @@ func (vlv VariableLoopingVUs) Run(ctx context.Context, out chan<- stats.SampleCo
 		currentMaxAllowedVUs = newMaxAllowedVUs
 	}
 
-	handleAllRawSteps := func() bool {
-		for {
+	// This is the step that is the actual gracefulStop at the end
+	lastStep := gracefulExecutionSteps[len(gracefulExecutionSteps)-1]
+	// TODO maybe instead of getting only the last step (above) mergeSteps should return the
+	// remaining gracefulExecutionSteps? or maybe we should refactor this even more ...
+	mergedSteps := mergeSteps(rawExecutionSteps, gracefulExecutionSteps[:len(gracefulExecutionSteps)-1])
+	timer := time.NewTimer(time.Hour * 24)
+	for _, step := range mergedSteps {
+		offsetDiff := step.step.TimeOffset - time.Since(startTime)
+		if offsetDiff > 0 { // wait until time of event arrives // TODO have a mininum
+			timer.Reset(offsetDiff)
 			select {
-			case step, ok := <-rawStepEvents:
-				if !ok {
-					return true
-				}
-				handleNewScheduledVUs(step.PlannedVUs)
-			case step := <-gracefulLimitEvents:
-				if step.PlannedVUs > currentMaxAllowedVUs {
-					// Handle the case where a value is read from the
-					// gracefulLimitEvents channel before rawStepEvents
-					handleNewScheduledVUs(step.PlannedVUs)
-				}
-				handleNewMaxAllowedVUs(step.PlannedVUs)
-			case <-ctx.Done():
-				return false
+			case <-maxDurationCtx.Done():
+				return // exit if context is cancelled
+			case <-timer.C:
+				// now we do a step
 			}
 		}
+		if step.raw {
+			handleNewScheduledVUs(step.step.PlannedVUs)
+		} else {
+			handleNewMaxAllowedVUs(step.step.PlannedVUs)
+		}
 	}
 
-	if handleAllRawSteps() {
-		// Handle any remaining graceful stops
-		go func() {
-			for {
-				select {
-				case step := <-gracefulLimitEvents:
-					handleNewMaxAllowedVUs(step.PlannedVUs)
-				case <-maxDurationCtx.Done():
-					return
-				}
+	go func() {
+		offsetDiff := lastStep.TimeOffset - time.Since(startTime)
+		if offsetDiff > 0 { // wait until time of event arrives // TODO have a mininum
+			timer.Reset(offsetDiff)
+			select {
+			case <-maxDurationCtx.Done():
+				return // exit if context is cancelled
+			case <-timer.C:
+				// now we do a step
 			}
-		}()
-	}
+		}
+		handleNewMaxAllowedVUs(lastStep.PlannedVUs)
```

Comment:
The last one is the one that is from the gracefulStop, as mentioned(in the commit message) previously all leftover gracefulStopSteps were handled differently in order for the gracefulStop to be implemented. 

I am of the opinion that this can be handled ... better in the sense that IMO most of the gracefulStop steps are redundant and possibly all of them, but the last oen. But for the purposes of getting 1007 out the door, this year, refactoring everything all the time is not the best plan IMO :D